### PR TITLE
Support Rust extension as a submodule in mixed Python/Rust projects

### LINKED ISSUE #475: How to create a mixed rust/python project with support of python's namespaces?
I need to create mixed rust/python project with python's namespace, named as `cykooz.resizer`. 

But `maturin` doesn't seem to support it. I has trying to set `lib.name` in `Cargo.toml` as `"cykooz.resizer.rust_lib"` but this hasn't worked. The command `maturin develop` has returned an error `"invalid character '.' in crate name: 'cykooz.resizer.rust_lib'"`.

Example of the layout I would like to get:

```
cykooz.resizer
├── Cargo.toml
├── cykooz
│   ├── __init__.py
│   └── resizer
│       ├── __init__.py
│       ├── bar.py
│       └── rust_lib.cpython-38-x86_64-linux-gnu.so
├── Readme.md
└── src
    └── lib.rs
```

### PR DESCRIPTION
**Pull request by messense**
By supporting dot in `package.metadata.maturin.name`.

Fixes #475

### CODE REVIEW

File: src/build_context.rs [hunk 1]
```diff
@@ -54,27 +54,59 @@ impl BridgeModel {
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum ProjectLayout {
     /// A rust crate compiled into a shared library with only some glue python for cffi
-    PureRust,
+    PureRust(String),
```
Comment (konstin) [inline src/build_context.rs:57]: What's the string in there (please add to doc comment)

File: src/build_context.rs [hunk 2]
```diff
@@ -54,27 +54,59 @@ impl BridgeModel {
...
     /// A python package that is extended by a native rust module.
     ///
     /// Contains the canonicialized (i.e. absolute) path to the python part of the project
-    Mixed(PathBuf),
+    Mixed {
```
Comment (konstin) [inline src/build_context.rs:61]: Also update this doc comment please; relative to what are those path names?

File: src/build_context.rs [hunk 3]
```diff
@@ -54,27 +54,59 @@ impl BridgeModel {
...
 impl ProjectLayout {
     /// Checks whether a python module exists besides Cargo.toml with the right name
     pub fn determine(project_root: impl AsRef<Path>, module_name: &str) -> Result<ProjectLayout> {
-        let python_package_dir = project_root.as_ref().join(module_name);
-        if python_package_dir.is_dir() {
-            if !python_package_dir.join("__init__.py").is_file() {
+        let parts: Vec<&str> = module_name.split('.').collect();
+        let (python_module, rust_module, extension_name) = if parts.len() > 1 {
+            let mut rust_module = project_root.as_ref().to_path_buf();
+            rust_module.extend(&parts[0..parts.len() - 1]);
```
Comment (konstin) [inline src/build_context.rs:74]: Isn't `parts[0..parts.len() - 1]` just the entire slice?

File: src/build_context.rs [hunk 4]
```diff
@@ -54,27 +54,59 @@ impl BridgeModel {
...
+    },
 }
 
 impl ProjectLayout {
     /// Checks whether a python module exists besides Cargo.toml with the right name
     pub fn determine(project_root: impl AsRef<Path>, module_name: &str) -> Result<ProjectLayout> {
-        let python_package_dir = project_root.as_ref().join(module_name);
-        if python_package_dir.is_dir() {
-            if !python_package_dir.join("__init__.py").is_file() {
+        let parts: Vec<&str> = module_name.split('.').collect();
```
Comment (konstin) [inline src/build_context.rs:71]:
```suggestion
        // A dot in the module name means the extension module goes into the module folder specified by the path
        let parts: Vec<&str> = module_name.split('.').collect();
```

File: src/metadata.rs [hunk 5]
```diff
@@ -112,14 +112,22 @@ impl Metadata21 {
             description = None;
             description_content_type = None;
         };
+        let name = extra_metadata
+            .name
+            .map(|name| {
+                if let Some(pos) = name.find('.') {
```
Comment (konstin) [inline src/metadata.rs:118]: Kinda sad `str::str_split_once` is still nightly only

File: src/build_context.rs (hunk 3 above)
Comment (messense) [reply src/build_context.rs:74]: No, `parts[0..=parts.len() - 1]` is the entire slice. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f48d54867e7a4879cf761036d71fab5d

**Review by konstin — APPROVED** [review]

**Pull request merged**